src/factor.h

```diff
@@ -63,10 +71,18 @@ class CompactFactor {
   inline double potential(const VariableInFactor vifs[],
                           const variable_value_t var_values[],
                           const variable_id_t vid,
-                          const variable_value_t proposal) const {
+                          const variable_value_t proposal,
+                          feature_value_t val = DEFAULT_FEATURE_VALUE) const {
+    // For boolean, this is stored in CompactFactor.value;
+    //     caller shouldn't set the val arg.
+    // For categorical, this is stored in Factor.factor_params.value.
+    //     caller should pass in via the val arg.
+    // TODO: better data structures than the factor_params duct tape...
+    // See also: CompactFactorGraph::potential
+    val = (func_id == FUNC_AND_CATEGORICAL ? val : this->feature_value);
```

Comment:
Alright, then we should keep the param.  Let's just collapse the call stack such that `update_weight` directly calls this, and leave a comment on why we want the seemingly redundant param here.  I think it won't be too bad that way.